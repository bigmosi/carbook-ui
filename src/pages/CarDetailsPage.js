import { useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import { retrieveCar } from '../features/car/carSlice';
import '../styles/carDetails.css';

const CarDetailsPage = () => {
  const dispatch = useDispatch();
  const car = useSelector((state) => state.car.car);
  const dataFetchedRef = useRef(false);

  const { id } = useParams();
  useEffect(() => {
    if (dataFetchedRef.current) return;
    dataFetchedRef.current = true;
    dispatch(retrieveCar(id));
  }, [dispatch, id]);

  if (!car) {
    return <p>No car found</p>;
  }

  return (
    <div className="details">
      <div className="flex items-center h-screen">
        <div key={car.id} className="columns-1 md:columns-2 mx-8">
          <img className="w-full" src={car.image} alt="car" />
          <div className="mx-8 flex flex-col">
            <div className="justify-self-end">
              <h1 className="text-3xl font-bold">{car.brand}</h1>
              <h1>{car.description}</h1>
            </div>
            <div className="overflow-x-auto relative shadow-md sm:rounded-lg">
              <table className="w-full text-sm text-left text-gray-500 dark:text-gray-400">
                <tbody>
                  <tr className="bg-white border-b dark:bg-gray-900 dark:border-gray-700">
                    <th
                      scope="row"
                      className="py-4 px-6 font-medium text-gray-900 whitespace-nowrap dark:text-white"
                    >
                      Model
                    </th>
                    <td className="py-4 px-6">{car.model}</td>
                  </tr>
                  <tr className="bg-gray-50 border-b dark:bg-gray-800 dark:border-gray-700">
                    <th
                      scope="row"
                      className="py-4 px-6 font-medium text-gray-900 whitespace-nowrap dark:text-white"
                    >
                      Price
                    </th>
                    <td className="py-4 px-6">{car.price}</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div className="flex md:mt-24">
              <Link to={`/cars/${car.id}/reserve`}>
                <button
                  type="button"
                  className="btn btn-style justify-self-end rounded-full"
                >
                  Reserve car
                </button>
              </Link>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default CarDetailsPage;
