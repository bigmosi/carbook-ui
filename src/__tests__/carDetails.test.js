import React from 'react';
import '@testing-library/jest-dom';
import renderWithProviders from '../utils/test-utils';
import CarDetailsPage from '../pages/CarDetailsPage';

describe('CarDetails component', () => {
  const { asFragment } = renderWithProviders(<CarDetailsPage />);
  const firstRender = asFragment();

  it('should render correctly', () => {
    expect(firstRender).toMatchSnapshot();
  });
  it('should render correctly after state change', () => {
    expect(asFragment()).toMatchSnapshot();
  });
});
