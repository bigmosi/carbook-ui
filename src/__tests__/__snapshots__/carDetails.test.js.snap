// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarDetails component should render correctly 1`] = `
<DocumentFragment>
  <p>
    No car found
  </p>
</DocumentFragment>
`;

exports[`CarDetails component should render correctly after state change 1`] = `<DocumentFragment />`;
